microsim/opencl/ramp/kernels/ramp_ua.cl

```diff
@@ -0,0 +1,315 @@
+}
+
+// Given their current status, accumulate hazard from each person into their candidate places.
+kernel void people_send_hazards(uint npeople,
+                                uint nslots,
+                                global const uint* people_statuses,
+                                global const uint* people_place_ids,
+                                global const float* people_flows,
+                                global const float* people_hazards,
+                                volatile global uint* place_hazards,
+                                volatile global uint* place_counts,
+                                global const uint* place_activities,
+                                global const Params* params) {
+  int person_id = get_global_id(0);
+  if (person_id >= npeople) return;
+
+  // Early return for non infectious people
+  DiseaseStatus person_status = (DiseaseStatus)people_statuses[person_id];
+  if (!is_infectious(person_status)) return;
+
+  for (int slot=0; slot < nslots; slot++) {
+    // Get the place and flow for this slot
+    uint flow_idx = get_1d_index(person_id, slot, nslots);
+    uint place_id = people_place_ids[flow_idx];
+
+    //check it is not an empty slot
+    if (place_id == sentinel_value) continue;
+
+    float flow = people_flows[flow_idx];
+    uint activity = place_activities[place_id];
+    //check it is a valid activity and select hazard multiplier
+    flow *= (0 <= activity && activity <= 4) ? params->place_hazard_multipliers[activity] : 1.0;
+
+    // Convert the flow to fixed point
+    uint fixed_flow = (uint)(fixed_factor * flow);
+
+    // Atomically increment hazards and counts for this place
+    atomic_add(&place_hazards[place_id], fixed_flow);
+    atomic_add(&place_counts[place_id], 1);
+  }
+}
+
+//For each person accumulate hazard from all the places stored in their slots.
+kernel void people_recv_hazards(uint npeople,
+                                uint nslots,
+                                global const uint* people_statuses,
+                                global const uint* people_place_ids,
+                                global const float* people_flows,
+                                global float* people_hazards,
+                                global const uint* place_hazards,
+                                global const Params* params) {
+  int person_id = get_global_id(0);
+  if (person_id >= npeople) return;
+
+  // Early return for non susceptible people
+  DiseaseStatus person_status = (DiseaseStatus)people_statuses[person_id];
+  if (person_status != Susceptible) return;
+
+  // Initialize hazard to accumulate into
+  float hazard = 0.0;
+
+  for (int slot=0; slot < nslots; slot++) {
+    // Get the place and flow for this slot
+    uint flow_idx = get_1d_index(person_id, slot, nslots);
+    uint place_id = people_place_ids[flow_idx];
+    
+    //check it is not an empty slot
+    if (place_id == sentinel_value) continue;

```

Comment:
I'm aiming to avoid changing the OpenCL kernel code in this PR, to make it easier to assert that no functionality has changed, perhaps we can create a ticket to remind ourselves to do this in a subsequent PR?

Reply:
Yep, good call.